### PR TITLE
[Op][Dist] Add alltoall communication op

### PR DESCRIPTION
**Pull request by chenyu-jiang**
## Description ##
This PR implements a basic alltoall communication op.

Input tensor is evenly split into n parts (n is the world size) at axis 0. Part i of the tensor is sent to rank i. This op also supports grouping the alltoall of multiple tensors (by using a list of tensors as input).

## Checklist ##

- [x] PR's title starts with a category (e.g. [BUGFIX], [MODEL], [TUTORIAL], [FEATURE], [DOC], etc)
- [x] Changes are complete (i.e. I finished coding on this PR)
- [x] All changes have test coverage
- [x] Code is well-documented

cc @awslabs/raf-reviewer @zhen-jia

### CODE REVIEW

**Comment by yetiansh** [comment]
I'm not sure which name is better for this op, all_to_all or alltoall?

**Comment by zhen-jia** [comment]
Maybe `all_to_all` is better. What are the exact names used in other framework @StrongSpoon?

**Comment by comaniac** [comment]
Let's use `all_to_all`, following the NCCL naming convention: https://docs.nvidia.com/deeplearning/nccl/user-guide/docs/usage/p2p.html#all-to-all

**Comment by hgt312** [comment]
why there is grad for this op, i think dist op should not have grad

**Comment by yetiansh** [comment]
> why there is grad for this op, i think dist op should not have grad

I think `all_to_all` ops already exist in distributed MoE models before we apply `AutoGrad` pass, therefore they would have corresponding grads. The same `all_to_all` operations are applied to these grads to make results correct.

**Comment by chenyu-jiang** [comment]
Thanks for the suggestions! I have renamed the op to all_to_all.

File: python/raf/distributed/op.py [hunk 1]
```diff
@@ -193,6 +193,29 @@ def broadcast(x, root):
     return sym._broadcast(x, root)
 
 
+def all_to_all(x):
+    """Performs an all-to-all communication across all ranks.
+
+    Parameters
+    ----------
+    x : Tensor | List[Tensor]
```
Comment (comaniac) [inline python/raf/distributed/op.py:201]:
```suggestion
    x : Tensor or List[Tensor]
```

File: python/raf/distributed/op.py [hunk 2]
```diff
@@ -193,6 +193,29 @@ def broadcast(x, root):
...
+        The tensor(s) to perform all-to-all on. The input tensor is evenly split
+        into n chunks at axis 0, and chunk[i] is sent to rank i. If the input is a
+        list of tensors, the result is equivalent to calling all-to-all on each tensor
+        individually, but only one fused all-to-all call will be launched.
+
+    Returns
+    -------
+    ret: Tensor | List[Tensor]
```
Comment (comaniac) [inline python/raf/distributed/op.py:209]:
```suggestion
    ret: Tensor or List[Tensor]
```

**Review by comaniac — APPROVED** [review]
LGTM. Just nits.
cc @zhen-jia @hgt312

File: src/op/dialect/nccl/nccl.cc [hunk 3]
```diff
@@ -505,6 +505,150 @@ class NCCLBroadcast : public NCCLOpEnv {
...
+      for (int i = 0; i < input_x->fields.size(); ++i) {
+        DLTensor* x = input_x->fields[i];
+        size_t size_per_rank = tuple_sizes[i] / nccl_num_ranks;
+        for (int j = 0; j < nccl_num_ranks; ++j) {
+          void* in = reinterpret_cast<uint8_t*>(in_buffer) + offset + j * itvl;
+          void* x_ = reinterpret_cast<uint8_t*>(x->data) + j * size_per_rank;
+          CUDA_CALL(cudaMemcpyAsync(in, x_, size_per_rank, cudaMemcpyDeviceToDevice,
+                                    (cudaStream_t)stream));
+        }
+        offset += size_per_rank;
```
Comment (zhen-jia) [inline src/op/dialect/nccl/nccl.cc:608]: I think the alternative of memory copy is just launching more `send` and `receive`. If we use NCCL group call, is there a situation that memory copy cause more overhead, even though memory copy is assigned to another stream?

File: src/op/declare/collective_comm.cc [hunk 4]
```diff
@@ -208,6 +208,33 @@ void Send(const CallValues& call) {
...
+  const DLTensor* x = tv[0];
+  call->device = x->device;
+  for (int i = 0; i < tv.size(); ++i) {
+    const DLTensor* x = tv[i];
+    std::vector<int64_t> shape(x->shape, x->shape + x->ndim);
+    ret.push_back(TensorValue::Assemble(/*dev=*/x->device,
+                                        /*dtype=*/x->dtype,
+                                        /*shape=*/shape));
+  }
+  if (ret.size() == 0) {
```
[the remark this answers is not in the document]
Comment (chenyu-jiang) [reply src/op/declare/collective_comm.cc:225]: I can't recall a situation where this could happen. @yetiansh Do you remember why we added this line?

File: src/op/dialect/nccl/nccl.cc (hunk 3 above)
Comment (chenyu-jiang) [reply src/op/dialect/nccl/nccl.cc:608]: Indeed. We can try to do some profile and see which is better. I guess it depends on message size etc.

File: src/op/declare/collective_comm.cc (hunk 4 above)
[the remark this answers is not in the document]
Comment (yetiansh) [reply src/op/declare/collective_comm.cc:225]: I think I added this to follow the declaration of `AllReduce`. It's fine to remove this situation.

File: src/op/dialect/nccl/nccl.cc (hunk 3 above)
Comment (zhen-jia) [reply src/op/dialect/nccl/nccl.cc:608]: If so, could we set another parameter to decide whether to perform memory copy? The default value is beneficial for most cases.

**Comment by comaniac** [comment]
Can anyone confirm the status of this PR?

**Comment by zhen-jia** [comment]
@chenyu-jiang could you make a plan for this PR?

File: src/op/declare/collective_comm.cc (hunk 4 above)
[the remark this answers is not in the document]
Comment (chenyu-jiang) [reply src/op/declare/collective_comm.cc]: The condition is removed.

**Comment by chenyu-jiang** [comment]
Sorry I haven't run the benchmark yet. I will do the profile to find out whether we should use memory copy these two days. For now, I added a parameter for user to set whether memcpy should be used. We can remove the parameter if the benchmark result shows one method is strictly better than the other.

**Comment by chenyu-jiang** [comment]
Hi, I've got some benchmark results on this issue. The comparison was done using two settings: 1. all_to_all within a single p4d node, 2. all_to_all between two p4d nodes.


**TL;DR:** use direct send & recv (`group_use_memcpy=False`) intra-node. Set it to true may help when there is cross-machine communication and the number of groups is large.


###  Single p4d performance
image

In the benchmark, we test the performance under different input sizes and the number of groups (i.e. the length of the input tensor list). The x axis of the above figure correspond to the data size sent by a single GPU in total (i.e. summation of all groups).

On a single p4d node, directly launching more `ncclSend` and `ncclRecv` is always better than using memory copy. p4d connects all GPUs to a NVSwitch, so intra-node all_to_all is very fast. In this case, the memory copy causes a relatively large overhead.

### Cross 2 p4d nodes
image

However it is different when there's cross-node communication. On 2 p4d nodes, using memory copy is comparable to directly use `ncclSend` and `ncclRecv` when the number of groups is 2, and is better for most other cases (although the speedup is not very significant). In this case it could be that the small cross machine sends and recvs have a relatively high latency themselves, and manually grouping them using memcpy improves bandwidth utilization on the IB link between the machines.

Therefore I think we may keep the current implementation (have `group_use_memcpy` default to `false`), as it performs OK for most of the cases. User can choose to set to to true if they feel the latency of small sends/recvs is high.

**Comment by chenyu-jiang** [comment]
@zhen-jia Could you take another look? Thanks.

**Comment by zhen-jia** [comment]
Thanks @chenyu-jiang and @comaniac
